### DISCUSSION TITLE
Migratior up function throws error when running on Bun

### DISCUSSION THREAD

**Discussion by danlannz** [mikro-orm/mikro-orm#5349] [General]
When running migratior.up() on bun I'm getting migration error as MigrationClass is undefined.

```js
MigrationError: Migration Migration20240318042632 (up) failed: Original error: undefined is not a constructor (evaluating 'new (Object.values(migration))[0](this.driver, this.config)')
 info: {
  "direction": "up",
  "name": "Migration20240318042632",
  "path": "/mikro-orm-elysia/src/migrations/Migration20240318042632.ts",
  "context": {}
}

      at new ErrorWithCause (/mikro-orm-elysia/node_modules/pony-cause/lib/error-with-cause.js:4:12)
      at new MigrationError (/mikro-orm-elysia/node_modules/umzug/lib/umzug.js:45:9)
      at /mikro-orm-elysia/node_modules/umzug/lib/umzug.js:177:31

196 |     }
197 |     resolve(params) {
198 |         const createMigrationHandler = async (method) => {
199 |             const migration = await core_1.Utils.dynamicImport(params.path);
200 |             const MigrationClass = Object.values(migration)[0];
201 |             const instance = new MigrationClass(this.driver, this.config);
                                                       ^
TypeError: undefined is not a constructor (evaluating 'new (Object.values(migration))[0](this.driver, this.config)')
      at /mikro-orm-elysia/node_modules/@mikro-orm/migrations/Migrator.js:201:50
```

Doing some debugging it appears that the Object.values returns an empty array from the dynamic import object on Bun
https://github.com/mikro-orm/mikro-orm/blob/8799461ddf05637c1d227e3c7879c99b9bb74bd2/packages/migrations/src/Migrator.ts#L250-L251

I'm currently using the following workaround, will open a PR if this workaround is also compatible with all supported runtimes
```ts
  const migration = await Utils.dynamicImport(params.path!); 
  const fileName = params.name.split('.')[0];
  const MigrationClass = migration[fileName];
```

**Comment by B4nan** [comment]
Bugs in bun should be reported to bun, they claim to be node compatible, this is their problem to solve.

The proposed workaround won't work, the file name does not have to match the class name. Technically you could get around this by using `migrationsList`, handling the imports yourself:

https://mikro-orm.io/docs/migrations#importing-migrations-statically

**Reply by B4nan** [reply]
I got one more idea for a workaround, you could try to use `require` here - what the `Utils.dynamicImport` helper does is configurable via `dynamicImportProvider` ORM config option.

```ts
dynamicImportProvider: id => require(id),
```

**Reply by danlannz** [reply]
thanks @B4nan, I tried your suggested workaround but initially had the same issue.

Surprisingly, replacing `require(id)` with `import(id)` on your workaround did work, which I believe is the default implementation. This makes me think the issue is due to the way bun handles import() in the compiled CJS module. I will report the issue on the Bun repo 🙏

**Reply by B4nan** [reply]
Yes, its almost certainly something to fix on their end, there are quite a few open bug reports about dynamic import (but I haven't found one that would be matching this exactly, so better to have a new repro and open a new issue).

**Reply by B4nan** [reply]
Are you on windows btw? If so, maybe its about this bit:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L1062-L1068

**Reply by co-sic** [reply]
I'm on ubuntu and have the same problem for some time now. I had an ugly workaround in place, but now with the dynamicImportProvider this is at least a lot cleaner :)
